client/rpc/src/eth.rs

```diff
@@ -745,64 +745,105 @@ impl<B, C, P, CT, BE, H: ExHashT> EthApiT for EthApi<B, C, P, CT, BE, H> where
 	}
 
 	fn estimate_gas(&self, request: CallRequest, _: Option<BlockNumber>) -> Result<U256> {
-		let hash = self.client.info().best_hash;
-
-		let CallRequest {
-			from,
-			to,
-			gas_price,
-			gas,
-			value,
-			data,
-			nonce
-		} = request;
-
-		let gas_limit = gas.unwrap_or(U256::max_value()); // TODO: set a limit
-		let data = data.map(|d| d.0).unwrap_or_default();
-
-		let used_gas = match to {
-			Some(to) => {
-				let info = self.client.runtime_api()
-					.call(
-						&BlockId::Hash(hash),
-						from.unwrap_or_default(),
-						to,
-						data,
-						value.unwrap_or_default(),
-						gas_limit,
-						gas_price,
-						nonce,
-						true,
-					)
-					.map_err(|err| internal_err(format!("runtime error: {:?}", err)))?
-					.map_err(|err| internal_err(format!("execution fatal: {:?}", err)))?;
-
-				error_on_execution_failure(&info.exit_reason, &info.value)?;
-
-				info.used_gas
-			},
-			None => {
-				let info = self.client.runtime_api()
-					.create(
-						&BlockId::Hash(hash),
-						from.unwrap_or_default(),
-						data,
-						value.unwrap_or_default(),
-						gas_limit,
-						gas_price,
-						nonce,
-						true,
-					)
-					.map_err(|err| internal_err(format!("runtime error: {:?}", err)))?
-					.map_err(|err| internal_err(format!("execution fatal: {:?}", err)))?;
-
-				error_on_execution_failure(&info.exit_reason, &[])?;
+		let calculate_gas_used = |request| {
+			let hash = self.client.info().best_hash;
+
+			let CallRequest {
+				from,
+				to,
+				gas_price,
+				gas,
+				value,
+				data,
+				nonce
+			} = request;
+
+			let gas_limit = gas.unwrap_or(U256::max_value()); // TODO: set a limit
+			let data = data.map(|d| d.0).unwrap_or_default();
+
+			let used_gas = match to {
+				Some(to) => {
+					let info = self.client.runtime_api()
+						.call(
+							&BlockId::Hash(hash),
+							from.unwrap_or_default(),
+							to,
+							data,
+							value.unwrap_or_default(),
+							gas_limit,
+							gas_price,
+							nonce,
+							true,
+						)
+						.map_err(|err| internal_err(format!("runtime error: {:?}", err)))?
+						.map_err(|err| internal_err(format!("execution fatal: {:?}", err)))?;
+
+					error_on_execution_failure(&info.exit_reason, &info.value)?;
+
+					info.used_gas
+				},
+				None => {
+					let info = self.client.runtime_api()
+						.create(
+							&BlockId::Hash(hash),
+							from.unwrap_or_default(),
+							data,
+							value.unwrap_or_default(),
+							gas_limit,
+							gas_price,
+							nonce,
+							true,
+						)
+						.map_err(|err| internal_err(format!("runtime error: {:?}", err)))?
+						.map_err(|err| internal_err(format!("execution fatal: {:?}", err)))?;
+
+					error_on_execution_failure(&info.exit_reason, &[])?;
+
+					info.used_gas
+				},
+			};
 
-				info.used_gas
-			},
+			Ok(used_gas)
 		};
+		if cfg!(not(feature = "rpc_binary_search_estimate")) {
```

Comment:
A small grumble -- use `cfg!(feature = "rpc_binary_search_estimate")` without the `not` and reverse the if statement.